New blog post about configuring git repositories (#24)
eclipse-che/blog: _posts/2022-XX-XX-configuring-git-repositories.adoc

I recommend to first fork the repository, so we can later try commit and push our changes.

.Fork of this blog's repository on GitHub
image::/assets/img/configuring-git-repositories/blog-github-fork.png[Fork of blog repository]
{empty}

In this repo we can find a file called `devfile.yaml` containing specification for our workspace. You should know about the meaning of this file from previous blog posts.

Now just simply copy the git repo URL, paste it in the `Import from Git` part of {prod-short} dashboard and click `Create & Open`.

.Importing a workspace from Git
image::/assets/img/configuring-git-repositories/importing-from-git.png[Importing from Git]
{empty}

{prod-short} should first find the devfile in your repository and then begin to create a workspace based on it.

.Workspace being created
image::/assets/img/configuring-git-repositories/workspace-starting.png[Workspace creation]
{empty}

After successful creation of the workspace, you should be able to read and change the blog's content in {prod} Theia IDE. Accessing public repository from GitLab or Bitbucket would include the same steps.

> .Blog workspace opened in Theia

Author: I would like to simplify this title a bit.

> .Blog workspace opened in Theia

Reviewer: How about the following?
```suggestion
.New workspace with the repository files accessible in Che-Theia
```

image::/assets/img/configuring-git-repositories/git-config.png[Git config]
{empty}

Now the only thing left to do is commit your changes and push them to a remote repository.

.Push to the test branch
image::/assets/img/configuring-git-repositories/source-control.png[Push to test branch]
{empty}

From here you can create pull request to the https://github.com/eclipse-che/blog[eclipse-che/blog] repository or continue however you like.

=== Accessing a private Git repository

There is some action required both from you as a non-privileged user and from your {prod-short} admin in order to access private Git repository. We are going to configure OAuth authorization so that {prod-short} can act on behalf of your account on SCM server. GitHub and GitLab use OAuth2, while Bitbucket is a little behind, still using OAuth1.

==== User side
First, you'll need to create reference to {prod-short} in your SCM server account. This reference is called:

- https://docs.github.com/en/developers/apps/building-oauth-apps/creating-an-oauth-app[OAuth App] (GitHub)

- https://docs.gitlab.com/ee/integration/oauth_provider.html#authorized-applications[Authorized Application] (GitLab)

> - https://confluence.atlassian.com/adminjiraserver/using-applinks-to-link-to-other-applications-938846918.html[Application Link] (Bitbucket).

Reviewer: ```suggestion
- https://confluence.atlassian.com/adminjiraserver/using-applinks-to-link-to-other-applications-938846918.html[Application Link] (Bitbucket)
```

===== Creating Bitbucket Application Link

Creating the reference to {prod-short} on Bitbucket is a little more cumbersome.

First, we'll create RSA key-pair:
----
openssl genrsa -out private.pem 2048
openssl rsa -in private.pem -pubout > public.pub
openssl pkcs8 -topk8 -inform pem -outform pem -nocrypt -in private.pem -out privatepkcs8.pem
cat public.pub | sed 's/-----BEGIN PUBLIC KEY-----//g' | sed 's/-----END PUBLIC KEY-----//g' | tr -d '\n' > public-stripped.pub
cat privatepkcs8.pem | sed 's/-----BEGIN PRIVATE KEY-----//g' |  sed 's/-----END PRIVATE KEY-----//g' | tr -d '\n' > privatepkcs8-stripped.pem
----

These commands result in creation of some intermediary files which we can ignore, but also _public-stripped.pub_ and _privatepkcs8-stripped.pem_ keys in the format we'll need.

Next, we'll have to create consumer key and shared secret:

----
openssl rand -base64 24 > bitbucket-consumer-key
openssl rand -base64 24 > bitbucket-shared-secret
----

Now we can go ahead and create the Application link.

> Type {prod-short} URL as the *Application URL*. The default value is \https://che-eclipse-che.__<domain>__/ and click *[ Create new link ]*.

Author: Note to myself to generalize the URL.

In the warning stating _"The supplied Application URL has redirected once"_ check the *Use this URL* checkbox and click *[ Continue ]*

Continue by filling the Application link form in this way:

* {prod-short} as *Application Name*.

* Generic Application as *Application Type*.

* {prod-short} as *Service Provider Name*.

* Paste the content of the `bitbucket-consumer-key` file as *Consumer key*.

* Paste the content of the `bitbucket-shared-secret` file as *Shared secret*.

* _<Bitbucket Server URL>_/plugins/servlet/oauth/request-token as *Request Token URL*.

* _<Bitbucket Server URL>_/plugins/servlet/oauth/access-token as *Access token URL*.

* _<Bitbucket Server URL>_/plugins/servlet/oauth/access-token as *Authorize URL*.

* Check the *Create incoming link* checkbox.

Click *[ Continue ]* and another smaller form should appear which you should fill out in this manner:

* Paste the content of the `bitbucket_consumer_key` file as *Consumer Key*

* {prod-short} as *Consumer name*

* Paste the content of the `public-stripped.pub` file as *Public Key*

And after clicking *[ Continue ]* button you should be all set. You can now continue with the admin section.

==== Admin side

> After you've created connection on the git service provider side, your {prod-short} admin will need to create following Kubernetes secret.

Reviewer: ```suggestion
After you've created the reference on the Git SCM, your {prod-short} administrator will create the following Kubernetes secret.
```

> ---

Author: Is the date somewhere specified or is it up to me? The schedule in https://github.com/eclipse-che/blog/issues/19 is oudated.

> ---

Reviewer 2: I think you can pick the date and update the issue description

Creating the reference on GitLab is similar to how it's done on GitHub. Additionally, you must choose the allowed scopes.

* Enter *{prod-short}* as the *Name*.

* Enter the {prod-short} OAuth callback URL as the *Redirect URI*. The default value is \https://{prod-deployment}-{prod-namespace}.__<domain>__/api/oauth/callback.

* Check the *Confidential* and *Expire* access tokens checkboxes.

* Under *Scopes*, check the `api`, `write_repository`, and `openid` checkboxes.

Click *Save application* to create the Authorized Application.

Copy and save the *Application ID* and *Secret* values because you will provide them to your {prod-short} administrator.

===== Creating the Bitbucket Application Link

Create an RSA key-value pair:
Use a command line to generate the *public-stripped.pub* and *privatepkcs8-stripped.pem* keys in the required format:
----
$ openssl genrsa -out private.pem 2048

$ openssl rsa -in private.pem -pubout > public.pub

$ openssl pkcs8 -topk8 -inform pem -outform pem -nocrypt -in private.pem -out privatepkcs8.pem

$ cat public.pub | sed 's/-----BEGIN PUBLIC KEY-----//g' | sed 's/-----END PUBLIC KEY-----//g' | tr -d '\n' > public-stripped.pub

> $ cat privatepkcs8.pem | sed 's/-----BEGIN PRIVATE KEY-----//g' |  sed 's/-----END PRIVATE KEY-----//g' | tr -d '\n' > privatepkcs8-stripped.pem

Author: No it's not here. But it's not only the private key. The consumer key also needs to be base64 encoded (even though it's created like this `openssl rand -base64 24`). And I noticed something that's not in our documentation (https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-authorization/#proc_configuring-bitbucket-server-oauth1_che), but is in our testing script. The bitbucket secret also needs to contain the shared secret. I am going to update it.

> $ cat privatepkcs8.pem | sed 's/-----BEGIN PRIVATE KEY-----//g' |  sed 's/-----END PRIVATE KEY-----//g' | tr -d '\n' > privatepkcs8-stripped.pem

Reviewer: @xbaran4, by "_something that's not in our documentation_" do you mean https://github.com/skabashnyuk/gitsrv/blob/main/bitbucket/4_setupche.sh ?

> $ cat privatepkcs8.pem | sed 's/-----BEGIN PRIVATE KEY-----//g' |  sed 's/-----END PRIVATE KEY-----//g' | tr -d '\n' > privatepkcs8-stripped.pem

Reviewer: @xbaran4, can I add `base64` here then?
 
```suggestion
$ cat privatepkcs8.pem | sed 's/-----BEGIN PRIVATE KEY-----//g' |  sed 's/-----END PRIVATE KEY-----//g' | tr -d '\n' | base64 > privatepkcs8-stripped.pem
```

Enter the following details for the *Application* link:

* Enter *{prod-short}* as the *Application Name*.

* Select *Generic Application* as the *Application Type*.

* Enter *{prod-short}* as the *Service Provider Name*.

* Paste the content of the *bitbucket-consumer-key* file as the *Consumer key*.

* Paste the content of the *bitbucket-shared-secret* file as the *Shared secret*.

* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/request-token` as the *Request Token URL*.

* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/access-token` as the *Access token URL*.

* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/access-token` as the *Authorize URL*.

* Check the *Create incoming link* checkbox.

Click *Continue* and enter the following details:

* Paste the content of the *bitbucket_consumer_key* file as the *Consumer Key*.

* Enter *{prod-short}* as the *Consumer name*.

* Paste the content of the *public-stripped.pub* file as the *Public Key*.

Click *Continue*.

This completes your user steps. Additional steps are now required from your {prod-short} administrator.

==== Admin side

> After you've created the reference on the Git SCM, your {prod-short} administrator will create the following Kubernetes secret.

Reviewer: Let's say I am a Che user. Can I have my own private fork of our team's private project repository. If I set up, let's say a GitHub OAuth App, and the Che admin applies a Secret for it. Will this Secret enable my colleagues to access my private (my fork) repository?

Author: You need to go through the oauth flow to get an access token (which is then used to read/write etc). This access token is created for your Che user in his k8s namespace. If your colleague wanted his own access token, he would need to go through the oauth flow as well. Which means he would need to login to GitHub with your GitHub credentials.

<2> The {prod-short} namespace. The default is `{prod-namespace}`.
<3> Depending on your choice of the Git SCM service provider: `github` or `gitlab`.
<4> The GitLab server URL (only for GitLab).
<5> Base64 encoded: the GitHub OAuth Client ID or GitLab Application ID.
<6> Base64-encoded: the GitHub OAuth Client Secret or GitLab Client Secret.

==== Bitbucket

A secret for Bitbucket differs from the GitHub and GitLab secrets mainly in the `data` section.

The administrator can apply the secret by using `{orch-cli}` on a command line:

[source, bash]
----
{orch-cli} apply -f - <<EOF
kind: Secret
apiVersion: v1
metadata:
  name: bitbucket-oauth-config
  namespace: __<...>__ <1>
  labels:
    app.kubernetes.io/component: oauth-scm-configuration
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:
    che.eclipse.org/oauth-scm-server: bitbucket
    che.eclipse.org/scm-server-endpoint: __<...>__ <2>
type: Opaque
data:
  private.key: __<...>__ <3>
  consumer.key: __<...>__ <4>
EOF
----
<1> The {prod-short} namespace. The default is `{prod-namespace}`.
<2> The Bitbucket server URL.
<3> Base64-encoded content of the `privatepkcs8-stripped.pem` file.

> <4> Base64-encoded content of the `bitbucket_server_consumer_key` file.

Reviewer: @xbaran4, I can't find any other mention of `bitbucket_server_consumer_key` in this file. Could it be that `bitbucket_consumer_key` is meant here?
```suggestion
<4> Base64-encoded content of the `bitbucket_consumer_key` file.
```

Author: Yes, you are right. Good catch!

.Source control options in Che-Theia
image::/assets/img/configuring-git-repositories/source-control.png[Source control in Che-Theia]
{empty}

I must also set my Git user name and user email (if I haven't done so already) that will identify my commits.

.Configuring Git user name and user email in Che-Theia
image::/assets/img/configuring-git-repositories/git-config.png[Git config in Che-Theia]
{empty}

 I commit my changes and push them to the remote Git repository.

.Using the Push option in Che-Theia to push to the test branch
image::/assets/img/configuring-git-repositories/source-control.png[Using the Push option in Che-Theia]
{empty}

From here, I can open a pull request to the https://github.com/eclipse-che/blog[eclipse-che/blog] repository.

=== Accessing a private Git repository

Before you can access a private Git repository, some actions are required from you as a non-privileged user and from your {prod-short} administrator. Those actions are required for configuring the OAuth authorization so that {prod-short} can act on behalf of the user account on the Git SCM server. GitHub and GitLab use OAuth2, whereas Bitbucket uses OAuth1.

> ==== User side

Reviewer 3: Hello, according to the [docs](https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-authorization/#configuring-github-oauth_che) and this comment: https://github.com/eclipse/che/issues/21298#issuecomment-1079020296  this should be an task meant to be performed by the admin.

Could we remove this `User side` section since the oauth app creation, retrieving client ID and secret should be also done by an admin?

----

Use a command line to generate a consumer key and a shared secret:

----
$ openssl rand -base64 24 > bitbucket-consumer-key

$ openssl rand -base64 24 > bitbucket-shared-secret
----

Create the *Application* link:

Enter the {prod-short} URL as the *Application URL*. The default value is `https://{prod-deployment}-{prod-namespace}.__<domain>__/ ` .

Click *Create new link*.

Under *The supplied Application URL has redirected once*, check the *Use this URL* checkbox and click *Continue*.

Enter the following details for the *Application* link:

* Enter *{prod-short}* as the *Application Name*.

* Select *Generic Application* as the *Application Type*.

* Enter *{prod-short}* as the *Service Provider Name*.

* Paste the content of the *bitbucket-consumer-key* file as the *Consumer key*.

* Paste the content of the *bitbucket-shared-secret* file as the *Shared secret*.

* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/request-token` as the *Request Token URL*.

* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/access-token` as the *Access token URL*.

> * Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/access-token` as the *Authorize URL*.

Reviewer 3: ```suggestion
* Enter `__<bitbucket_server_url>__/plugins/servlet/oauth/authorize` as the *Authorize URL*.
```

I think it should be `/plugins/servlet/oauth/authorize` like it is in the testing scripts? https://github.com/skabashnyuk/gitsrv/blob/248983e61ae20b47c11396b19ef923136f83ad1c/bitbucket/3_configure_links.sh#L27

Author: You are right, good eye. Looks like we had some mistakes in our docs https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-authorization/#proc_configuring-bitbucket-server-oauth1_che
